File: 백준/Silver/2304. 창고 다각형/창고 다각형.py
import sys
input = sys.stdin.readline

n = int(input())
li = []
bigl = 0
bigh = 0
for _ in range(n):
    l, h = map(int, input().split())
    li.append((l, h))
    if l>bigl:
        bigl = l
    if h>bigh:
        bigh = h
        idx = l

arr = [0] * (bigl+1)
for l, h in li:
    arr[l] = h

res = 0
tmp = 0
for i in range(idx+1):
    if arr[i] > tmp:
        tmp = arr[i]
    res += tmp

tmp = 0
for i in range(bigl, idx, -1):
    if arr[i]>tmp:
        tmp = arr[i]
    res += tmp

print(res)

File: 백준/Silver/2304. 창고 다각형/창고 다각형.java

import java.io.*;
import java.util.*;


public class Main {
	
	public static void main(String[] args) throws IOException {
		
		//1. 입력받고 L을 기준으로 정렬
		//2. 가장 높은 H 찾기
		//3. 가장 높은 H를 기준으로 왼쪽 오른쪽 탐색
		//4. 높은 H가 나오면 갱신, 아니면 그대로 면적 +
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int[][] arr = new int[n][2];
		
		for(int i = 0; i<n; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			for(int j = 0; j<2; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
			}
			
		}
		
		Arrays.sort(arr, (o1, o2) -> {
			return o1[0]-o2[0];
			});
		
		int idx = 0;
		int big_h = 0;
		for(int i = 0; i<n; i++) {
			if(arr[i][1]>big_h) {
				big_h = arr[i][1];
				idx = i;
			}
		}
		
		int res = 0;
		
		//왼쪽 탐색
		int tmp1 = 0;
		for(int i = 0; i<idx; i++) {
			if(arr[i][1]>tmp1) {
				tmp1 = arr[i][1];
			}
			res += (tmp1*(arr[i+1][0]-arr[i][0]));
		}
		
		//오른쪽 탐색
		int tmp2 = 0;
		for(int i = n-1; i>idx; i--) {
			if(arr[i][1]>tmp2) {
				tmp2 = arr[i][1];
			}
			res += (tmp2*(arr[i][0]-arr[i-1][0]));
		}
		
		res += big_h;
		System.out.println(res);
	
	}
}